linmod/models/utils.py

```diff
@@ -0,0 +1,221 @@
+import string
+from abc import ABC, abstractmethod
+from itertools import product
+from typing import Iterable
+
+import polars as pl
+from numpyro.diagnostics import summary
+from numpyro.infer.mcmc import MCMC
+from plotnine import aes, geom_line, ggplot, theme_bw
+
+
+class GeographicAggregator(ABC):
+    r"""
+    Aggregate forecasts from state level to some larger grouping, like HHS divisions.
+
+    __call__ arguments
+    forecast:          A LazyFrame with the standard model output format.
+    geo_map:           A dictionary mapping "divisions" in the forecast to new units.
+                       Every existing division need not be mapped.
+
+    Returns a LazyFrame in the standard output format on the new geographical units.
+    """
+
+    @abstractmethod
+    def __call__(
+        self, forecast: pl.LazyFrame, geo_map: dict[str, str], **kwargs
+    ) -> pl.LazyFrame:
+        raise NotImplementedError()
+
+
+class InfectionWeightedAggregator(GeographicAggregator):
+    r"""
+    Geographically aggregate while accounting for unequal numbers of infected individuals in different states.
+
+    For convenience, this is broken down into per-state total human populations, and the proportions of those populations
+    which are infected. The per-state weight in the aggregation is the product of these terms.
+
+    The state-level populations and proportions infected are taken to be constant over the forecasting horizon.
+
+    __call__ keyword arguments
+    pop_size:          A pl.LazyFrame with the (possibly relative) population size of each forecasted division.
+                       Must contain all divisions which are in the `geo_map`, and two columns: "division" and "pop_size".
+                       Defaults to equal population sizes.
+    prop_infected      A pl.LazyFrame with the proportion of the populations infected (or something proportionate thereto),
+                       taken to be a constant over the modeling period. Must contain all divisions which are in the `geo_map`,
+                       and two columns: "division" and "prop_infected". Defaults to equal proportions infected.
+    """
+
+    def __call__(
+        self, forecast: pl.LazyFrame, geo_map: dict[str, str], **kwargs
```

Comment:
geo_map should be annotated in the function doc?

Reply:
Good catch. Documentation here was split between parent and child classes.